README.md
# `snaps-cli`

A CLI for developing MetaMask Snaps, formerly known as "plugins."

## Compatibility

If you're looking for the `snaps-cli` compatible with the pre-2021 `metamask-snaps-beta`, please see [this branch](https://github.com/MetaMask/snaps-cli/tree/version-0.4.0).

## Installation

Use `node@^12.0.0`.
We recommend [nvm](https://github.com/nvm-sh/nvm) for managing Node versions.

- `yarn global add snaps-cli`

## Basic Usage

```bash
mkdir mySnap
cd mySnap
snap init
```

There is also an alias `mm-snap` in case of `snap` being taken.

## MetaMask Snaps

MetaMask Snaps consist of two things: a JSON manifest and a JavaScript bundle.
For a variety of reasons, we decided to follow NPM conventions. Thus, the manifest file
is simply `package.json`, with the following required standard fields:

- `name`
- `version`
- `description`
- `main` (relative path to source entry point, not the bundle)
- `repository`

In addition, we use the following, required custom fields:

- `web3Wallet` (`object`)
  - `bundle` (`object`)
    - `local` (`string`; relative path to bundle)
    - `url` (`string`; absolute URL to bundle)
      - Set to e.g. `localhost:8081/dist/bundle.js` for local development.
  - `initialPermissions` (`{ string: object }`; permissions to be requested on
    Snap installation)
    - See [this interface](https://github.com/MetaMask/rpc-cap#requestpermissions-irequestedpermissions)
    for details.

If you exclude any required fields from `package.json`, your Snap may not
work properly or install at all.

We recommend building your Snap using this tool.
You can bundle your Snap using your own tools, but it must run in the browser,
run in SES, and its contents must be wrapped in an anonymous function (`() => ( ... )`).

### Assumed Project Structure

This tool has default arguments assuming the following project structure:

```text
 .
 |- index.js
 |- dist/
 |-- bundle.js
 |- ... (all other project files and folders)
```

Source files other than `index.js` are located through its imports.
The defaults can be overwritten using the `snap.config.json` config file,
[see below](#configuration-file).

### Permissions

This module uses permissions as defined in [`rpc-cap`, MetaMask's JSON RPC
capabilities middleware](https://github.com/MetaMask/rpc-cap).
See examples in this repo for details.

## Usage

Always use `node@10.16.3`, the version currently used to develop MetaMask.

`snap --help`

```text
Usage: snap <command> [options]

Commands:
  snap init      Initialize Snap package                     [aliases: i]
  snap build     Build Snap from source                      [aliases: b]
  snap eval      Attempt to evaluate Snap bundle in SES      [aliases: e]
  snap manifest  Validate project package.json as a Snap manifest
                                                                    [aliases: m]
  snap serve     Locally serve Snap file(s) for testing      [aliases: s]
  snap watch     Build Snap on change                        [aliases: w]

Options (build):
  --help, -h                      Show help                            [boolean]
  --src, -s                       Source file
                                       [string] [required] [default: "index.js"]
  --dist, -d                      Output directory
                                          [string] [required] [default: "dist/"]
  --outfileName, -n               Output file name
                                                 [string] [default: "bundle.js"]
  --sourceMaps                    Whether building outputs sourcemaps
                                                      [boolean] [default: false]
  --stripComments, --strip        Whether to remove code comments from bundle
                                                      [boolean] [default: false]
  --port, -p                      Local server port for testing
                                             [number] [required] [default: 8081]
  --eval, -e                      Attempt to evaluate Snap bundle in SES
                                                       [boolean] [default: true]
  --manifest, -m                  Validate project package.json as a Snap
                                  manifest             [boolean] [default: true]
  --populate                      Update Snap manifest properties of
                                  package.json         [boolean] [default: true]
  --verboseErrors, -v, --verbose  Display original errors
                                                      [boolean] [default: false]
  --suppressWarnings, -w          Suppress warnings   [boolean] [default: false]

Examples:
  snap init                            Initialize Snap package from
                                            scratch
  snap build -s index.js -d out        Build 'Snap.js' as
                                            './out/bundle.js'
  snap build -s index.js -d out -n     Build 'Snap.js' as
  Snap.js                                 './out/Snap.js'
  snap serve -r out                    Serve files in './out' on port 8080
  snap serve -r out -p 9000            Serve files in './out' on port 9000
  snap watch -s index.js -d out        Rebuild './out/bundle.js' on changes
                                            to files in 'index.js' parent and
                                            child directories
```

### Usage Notes

- Commands
  - `watch --src ... --dist ...` rebuilds on all changes in the parent directory
    of `src` and its children except:
    - `node_modules/`
    - `test/` and `tests/`
    - The specified `dist` directory
    - Dotfiles
  - `serve --root ... --port ...` serves the `root` directory on `localhost:port`
    - By default, `root` serves the current working directory: `.`
- Arguments
  - `src`, `snap`, and `outfileName` must be file paths when specified
  - `dist` and `root` must be directory paths when specified

### Configuration File

`snap.config.json` can be placed in the project root directory. It should have string keys matching
command arguments. Values become argument defaults, which can still be overriden on the command line.
Example:

```json
{
  "src": "lib",
  "dist": "out",
  "port": 9000
}
```


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [0.4.2]
### Added
- API Compatability with latest Snaps implementation ([#114](https://github.com/MetaMask/snaps-cli/pull/114))
- License file ([#88](https://github.com/MetaMask/snaps-cli/pull/88))
- `--stripComments` option ([#65](https://github.com/MetaMask/snaps-cli/pull/65))

### Changed
- Use Node 14 ([#77](https://github.com/MetaMask/snaps-cli/pull/77))
- Prevent serve command from caching files ([#90](https://github.com/MetaMask/snaps-cli/pull/90))
- Restore crypto to eval environment ([#87](https://github.com/MetaMask/snaps-cli/pull/87))
- Fix `applyConfig` behavior ([#85](https://github.com/MetaMask/snaps-cli/pull/85), [#84](https://github.com/MetaMask/snaps-cli/pull/84))
- Alias for `suppressWarnings` to `--sw` ([#81](https://github.com/MetaMask/snaps-cli/pull/81))
- Use workers as the execution environment ([#60](https://github.com/MetaMask/snaps-cli/pull/60))
- Use `ses@^0.11.0` ([#60](https://github.com/MetaMask/snaps-cli/pull/60))

### Removed
- 3box example
- Environment option ([#86](https://github.com/MetaMask/snaps-cli/pull/86))

### Fixed
- File existence checks `init` command ([#73](https://github.com/MetaMask/snaps-cli/pull/73))
- `fs.exists` and `fs.readdir` error handling ([#71](https://github.com/MetaMask/snaps-cli/pull/71))

## [0.4.1]
### Uncategorized
- Rename `plugin` option to `bundle`, but keep alias
- Bugfixes after renaming

## [0.4.0]
### Uncategorized
- Rename package to `snaps-cli`
- Rename CLI entry point to `snap`
  - Maintain `mm-plugin` entry point alias and `mm-plugin.config.json` config files

## [0.3.13]
### Uncategorized
- Fix fatal `watch` command bug

## [0.3.12]
### Uncategorized
- Use existing `package.json` for `mm-plugin init`, if it exists
  - Fixes [#38](https://github.com/MetaMask/snaps-cli/issues/38)

## [0.3.11]
### Uncategorized
- Stop publishing example builds to `npm`
- Add `buildExamples` script to `package.json`
  - Use this to build all examples with source maps
- Rename `build` and `watch` option `debug` to `sourceMaps`
  - Maintain `debug` alias for backwards compatibility

## [0.3.10]
### Uncategorized
- Add optional source mapping via `debug` option to `build` and `watch` commands

## [0.3.9]
### Uncategorized
- Fix `undefined` `port` in `package.json` on build
- The `populate` option for `mm-plugin manifest` (and `mm-plugin build`, which calls
  `manifest` by default) no longer has the alias `p`
  - `p` is now in every case reserved as an alias for `port`.

## [0.3.8]
### Uncategorized
- Update readme

## [0.3.7]
### Uncategorized
- `mm-plugin manifest` now populates `package.json:web3Wallet.bundle.url` using config values
  - It only does this if `bundle.url` is missing or starts with `http://localhost`.
  - It applies: `bundle.url = 'http://localhost:${port}/${bundlePath}'`
- Basic handling of HTML comment syntax in bundle
  - `<!--` and `-->` can be valid JavaScript, but are always forbidden by SES.
    They are now destructed into `<! --` and `-- >`. This may break code in some edge cases.

## [0.3.6]
### Uncategorized
- Update faulty ignore files; package size decreased

## [0.3.5]
### Uncategorized
- Updated examples to work with `metamask-plugin-beta` as of [this commit](https://github.com/MetaMask/metamask-plugin-beta/commit/b8ba321689cec6749502969f0084e12193e92dab)

## [0.3.4]
### Uncategorized
- `mm-plugin watch` should no longer terminate on on parse or write errors

## [0.3.3]
### Uncategorized
- Update `realms-shim`

## [0.3.2]
### Uncategorized
- Handle SES edge cases
  - Babel: `regeneratorRuntime` global variable
  - Browserify: modules that use `Buffer`
    - Added regex that replaces lines in the bundle of the form `(function (Buffer){`

## [0.3.1]
### Uncategorized
- Rename `.mm-plugin.json` to `mm-plugin.config.json`
  - Still support `.mm-plugin.json` for backwards compatibility

## [0.3.0]
### Uncategorized
- Remove default command; at least one command must now be specified
- Add `init` command

## [0.2.1]
### Uncategorized
- Specifying `web3Wallet.bundle.local` or `dist` in `.mm-plugin.json` no longer
  overwrites the default for the `serve` command's `root` directory argument

## [0.2.0]
### Uncategorized
- Use named rather than positional arguments

[Unreleased]: https://github.com/MetaMask/snaps-cli/compare/v0.4.2...HEAD
[0.4.2]: https://github.com/MetaMask/snaps-cli/compare/v0.4.1...v0.4.2
[0.4.1]: https://github.com/MetaMask/snaps-cli/compare/v0.4.0...v0.4.1
[0.4.0]: https://github.com/MetaMask/snaps-cli/compare/v0.3.13...v0.4.0
[0.3.13]: https://github.com/MetaMask/snaps-cli/compare/v0.3.12...v0.3.13
[0.3.12]: https://github.com/MetaMask/snaps-cli/compare/v0.3.11...v0.3.12
[0.3.11]: https://github.com/MetaMask/snaps-cli/compare/v0.3.10...v0.3.11
[0.3.10]: https://github.com/MetaMask/snaps-cli/compare/v0.3.9...v0.3.10
[0.3.9]: https://github.com/MetaMask/snaps-cli/compare/v0.3.8...v0.3.9
[0.3.8]: https://github.com/MetaMask/snaps-cli/compare/v0.3.7...v0.3.8
[0.3.7]: https://github.com/MetaMask/snaps-cli/compare/v0.3.6...v0.3.7
[0.3.6]: https://github.com/MetaMask/snaps-cli/compare/v0.3.5...v0.3.6
[0.3.5]: https://github.com/MetaMask/snaps-cli/compare/v0.3.4...v0.3.5
[0.3.4]: https://github.com/MetaMask/snaps-cli/compare/v0.3.3...v0.3.4
[0.3.3]: https://github.com/MetaMask/snaps-cli/compare/v0.3.2...v0.3.3
[0.3.2]: https://github.com/MetaMask/snaps-cli/compare/v0.3.1...v0.3.2
[0.3.1]: https://github.com/MetaMask/snaps-cli/compare/v0.3.0...v0.3.1
[0.3.0]: https://github.com/MetaMask/snaps-cli/compare/v0.2.1...v0.3.0
[0.2.1]: https://github.com/MetaMask/snaps-cli/compare/v0.2.0...v0.2.1
[0.2.0]: https://github.com/MetaMask/snaps-cli/releases/tag/v0.2.0


snaps-cli.js
#!/usr/bin/env node

const { promises: fs } = require('fs');
const yargs = require('yargs');

const {
  build, init, manifest, snapEval, serve, watch,
} = require('./src/commands');

const { CONFIG_PATHS, logWarning } = require('./src/utils');

// globals

global.snaps = {
  verboseErrors: false,
  suppressWarnings: false,
  isWatching: false,
};

// yargs config and constants

const builders = {
  src: {
    alias: 's',
    describe: 'Source file',
    type: 'string',
    required: true,
    default: 'index.js',
  },

  dist: {
    alias: 'd',
    describe: 'Output directory',
    type: 'string',
    required: true,
    default: 'dist',
  },

  bundle: {
    alias: ['plugin', 'p', 'b'],
    describe: 'Snap bundle file',
    type: 'string',
    required: true,
    default: 'dist/bundle.js',
  },

  root: {
    alias: 'r',
    describe: 'Server root directory',
    type: 'string',
    required: true,
    default: '.',
  },

  port: {
    alias: 'p',
    describe: 'Local server port for testing',
    type: 'number',
    required: true,
    default: 8081,
  },

  sourceMaps: {
    alias: 'debug', // backwards compatibility
    describe: 'Whether builds include sourcemaps',
    type: 'boolean',
    default: false,
  },

  stripComments: {
    alias: 'strip',
    describe: 'Whether to remove code comments from the build output',
    type: 'boolean',
    default: false,
  },

  outfileName: {
    alias: 'n',
    describe: 'Output file name',
    type: 'string',
    default: 'bundle.js',
  },

  manifest: {
    alias: 'm',
    describe: 'Validate project package.json as a Snap manifest',
    type: 'boolean',
    default: true,
  },

  populate: {
    describe: 'Update Snap manifest properties of package.json',
    type: 'boolean',
    default: true,
  },

  eval: {
    alias: 'e',
    describe: `Attempt to evaluate Snap bundle in SES`,
    type: 'boolean',
    default: true,
  },

  verboseErrors: {
    alias: ['v', 'verbose'],
    type: 'boolean',
    describe: 'Display original errors',
    default: false,
  },

  suppressWarnings: {
    alias: ['w'],
    type: 'boolean',
    describe: 'Suppress warnings',
    default: false,
  },

  environment: {
    alias: ['env'],
    describe: 'The execution environment of the plugin.',
    type: 'string',
    default: 'worker',
    choices: ['worker'],
  },
};

main();

// application
async function main() {
  await applyConfig();

  // eslint-disable-next-line no-unused-expressions
  yargs
    .usage('Usage: $0 <command> [options]')
    .example('$0 init', `\tInitialize Snap package from scratch`)
    .example('$0 build -s index.js -d out', `\tBuild 'index.js' as './out/bundle.js'`)
    .example('$0 build -s index.js -d out -n snap.js', `\tBuild 'index.js' as './out/snap.js'`)
    .example('$0 serve -r out', `\tServe files in './out' on port 8080`)
    .example('$0 serve -r out -p 9000', `\tServe files in './out' on port 9000`)
    .example('$0 watch -s index.js -d out', `\tRebuild './out/bundle.js' on changes to files in 'index.js' parent and child directories`)

    .command(
      ['init', 'i'],
      'Initialize Snap package',
      (yarg) => {
        yarg
          .option('src', builders.src)
          .option('dist', builders.dist)
          .option('outfileName', builders.outfileName)
          .option('port', builders.port);
      },
      (argv) => init(argv),
    )

    .command(
      ['build', 'b'],
      'Build Snap from source',
      (yarg) => {
        yarg
          .option('src', builders.src)
          .option('dist', builders.dist)
          .option('outfileName', builders.outfileName)
          .option('sourceMaps', builders.sourceMaps)
          .option('stripComments', builders.stripComments)
          .option('port', builders.port)
          .option('eval', builders.eval)
          .option('manifest', builders.manifest)
          .option('populate', builders.populate)
          .option('environment', builders.environment)
          .implies('populate', 'manifest');
      },
      (argv) => build(argv),
    )

    .command(
      ['eval', 'e'],
      builders.eval.describe,
      (yarg) => {
        yarg
          .option('bundle', builders.bundle)
          .option('environment', builders.environment);
      },
      (argv) => snapEval(argv),
    )

    .command(
      ['manifest', 'm'],
      builders.manifest.describe,
      (yarg) => {
        yarg
          .option('dist', builders.dist)
          .option('port', builders.port)
          .option('populate', builders.populate);
      },
      (argv) => manifest(argv),
    )

    .command(
      ['serve', 's'],
      'Locally serve Snap file(s) for testing',
      (yarg) => {
        yarg
          .option('root', builders.root)
          .option('port', builders.port);
      },
      (argv) => serve(argv),
    )

    .command(
      ['watch', 'w'],
      'Build Snap on change',
      (yarg) => {
        yarg
          .option('src', builders.src)
          .option('dist', builders.dist)
          .option('outfileName', builders.outfileName)
          .option('sourceMaps', builders.sourceMaps)
          .option('environment', builders.environment)
          .option('stripComments', builders.stripComments);
      },
      (argv) => watch(argv),
    )

    .option('verboseErrors', builders.verboseErrors)
    .option('suppressWarnings', builders.suppressWarnings)
    .demandCommand(1, 'You must specify at least one command.')
    .strict()
    .middleware((argv) => {
      assignGlobals(argv);
      sanitizeInputs(argv);
    })
    .help()
    .alias('help', 'h')
    .fail((msg, err, _yargs) => {
      console.error(msg || err.message);
      if (err && err.stack && snaps.verboseErrors) {
        console.error(err.stack);
      }
      process.exit(1);
    })
    .argv;
}

// misc

function assignGlobals(argv) {
  if (['w', 'watch'].includes(argv._[0])) {
    snaps.isWatching = true;
  }
  if (Object.prototype.hasOwnProperty.call(argv, 'verboseErrors')) {
    snaps.verboseErrors = Boolean(argv.verboseErrors);
  }
  if (Object.prototype.hasOwnProperty.call(argv, 'suppressWarnings')) {
    snaps.suppressWarnings = Boolean(argv.suppressWarnings);
  }
}

/**
 * Sanitizes inputs. Currently:
 * - normalizes paths
 */
function sanitizeInputs(argv) {
  Object.keys(argv).forEach((key) => {
    if (typeof argv[key] === 'string') {
      if (argv[key] === './') {
        argv[key] = '.';
      } else if (argv[key].startsWith('./')) {
        argv[key] = argv[key].substring(2);
      }
    }
  });
}

/**
 * Attempts to read the config file and apply the config to
 * globals.
 */
async function applyConfig() {

  // first, attempt to read and apply config from package.json
  let pkg = {};
  try {
    pkg = JSON.parse(await fs.readFile('package.json'));

    if (pkg.main) {
      builders.src.default = pkg.main;
    }

    if (pkg.web3Wallet) {
      const { bundle } = pkg.web3Wallet;
      if (bundle && bundle.local) {
        const { local: bundlePath } = bundle;
        builders.bundle.default = bundlePath;
        let dist;
        if (bundlePath.indexOf('/') === -1) {
          dist = '.';
        } else {
          dist = bundlePath.substr(0, bundlePath.indexOf('/') + 1);
        }
        builders.dist.default = dist;
      }
    }
  } catch (err) {
    if (err.code !== 'ENOENT') {
      logWarning(`Warning: Could not parse package.json`, err);
    }
  }

  // second, attempt to read and apply config from config file,
  // which will always be preferred if it exists
  let cfg = {};
  for (const configPath of CONFIG_PATHS) {
    try {
      cfg = JSON.parse(await fs.readFile(configPath));
      break;
    } catch (err) {
      if (err.code !== 'ENOENT') {
        logWarning(`Warning: '${configPath}' exists but could not be parsed.`);
      }
    }
  }

  if (
    typeof cfg !== 'object' ||
    Object.keys(cfg).length === 0
  ) {
    return;
  }

  Object.keys(cfg).forEach((key) => {
    let k = key;
    // backwards compatibility
    if (k === 'verbose') {
      k = 'verboseErrors';
    } else if (k === 'debug') {
      k = 'sourceMaps';
    }

    builders[k].default = cfg[k];
  });
}
